include/class.thread.php

```diff
@@ -745,10 +745,20 @@ function postEmail($mailinfo) {
         if ($mailinfo['userId']
                 || strcasecmp($mailinfo['email'], $ticket->getEmail()) == 0) {
             $vars['message'] = $body;
-            $vars['userId'] = $mailinfo['userId'] ? $mailinfo['userId'] : $ticket->getUserId();
+            $vars['userId'] = $mailinfo['userId'] ?: $ticket->getUserId();
+            return $ticket->postMessage($vars, 'Email');
+        }
+        // Consider collaborator role (disambiguate staff members as
+        // collaborators)
+        elseif (($E = UserEmail::lookup($mailinfo['email']))
+            && ($C = Collaborator::lookup(array(
+                'ticketId' => $ticket->getId(), 'userId' => $E->user_id
+            )))
+        ) {
+            $vars['userId'] = $C->getUserId();
+            $vars['message'] = $body;
             return $ticket->postMessage($vars, 'Email');
         }
-        // XXX: Consider collaborator role
         elseif ($mailinfo['staffId']
```

Comment:
I think this should move up so that we check the $vars data before running a database query